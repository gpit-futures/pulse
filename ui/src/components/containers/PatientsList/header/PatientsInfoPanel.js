import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash/fp';
import { lifecycle } from 'recompose';
import { Row, Col } from 'react-bootstrap';

import PTButton from '../../../ui-elements/PTButton/PTButton';
import PTCustomCheckbox from './PTCustomCheckbox';
import { unmountOnBlur } from '../../../../utils/HOCs/unmount-on-blur.utils'

@lifecycle(unmountOnBlur)
export default class PatientsInfoPanel extends PureComponent {
  static propTypes = {
    onColumnsSelected: PropTypes.func.isRequired,
    selectedColumns: PropTypes.objectOf(PropTypes.bool).isRequired,
  };

  state = {
    isFilterInputVisible: false,
    isPatientInfoPanelVisible: false,
    selected: this.props.selectedColumns,
  };

  componentDidUpdate(prevProps, prevState) {
    /* istanbul ignore next */
    if (!_.isEqual(prevState.selected, this.state.selected)) this.props.onColumnsSelected(this.state.selected)
  }

  toggleCheckbox = /* istanbul ignore next */ key => this.setState((prevState) => {
    const newValue = !_.get(['selected', key])(prevState);
    return _.set(['selected', key], newValue)(prevState);
  });

  toggleMultipleCheckboxes = keys => () => {
    /* istanbul ignore next */
    const nextCheckedValues = keys.map(_.flow(_.pick(keys), _.every(_.eq(true)))(this.state.selected) ? _.stubFalse : _.stubTrue);
    this.setState(prevState => _.merge(prevState, { selected: _.zipObject(keys, nextCheckedValues) }));
  };

  render() {
    const { selected } = this.state;

    return (
      <div
        className="dropdown-menu dropdown-menu-panel dropdown-menu-right dropdown-menu-summary dropdown-menu-patients"
      >
        <div className="form-group-wrapper">
          <div>
            <div className="heading">PATIENT INFO</div>
            <div className="form-group">
              <Row>
                <PTCustomCheckbox title="Name" name="name" isChecked={selected.name} disabled />
                <PTCustomCheckbox title="Address" name="address" isChecked={selected.address} onChange={this.toggleCheckbox} />
                <PTCustomCheckbox title="Born" name="dateOfBirth" isChecked={selected.dateOfBirth} disabled />
                <PTCustomCheckbox title="Gender" name="gender" isChecked={selected.gender} disabled />
                <PTCustomCheckbox title="NHS No." name="id" isChecked={selected.id} onChange={this.toggleCheckbox} />
                <Col xs={6} sm={4}>
                  <PTButton className="btn btn-success btn-inverse btn-bold btn-smaller" onClick={this.toggleMultipleCheckboxes(['address', 'id'])}>
                    <span className="btn-text">Select All</span>
                  </PTButton>
                </Col>
              </Row>
            </div>
          </div>
          <div>
            <div className="heading">DATE / TIME</div>
            <div className="form-group">
              <Row>
                <PTCustomCheckbox title="Orders" name="ordersDate" isChecked={selected.ordersDate} onChange={this.toggleCheckbox} />
                <PTCustomCheckbox title="Results" name="resultsDate" isChecked={selected.resultsDate} onChange={this.toggleCheckbox} />
                <PTCustomCheckbox title="Vitals" name="vitalsDate" isChecked={selected.vitalsDate} onChange={this.toggleCheckbox} />
                <PTCustomCheckbox title="Diagnosis" name="diagnosesDate" isChecked={selected.diagnosesDate} onChange={this.toggleCheckbox} />
                <PTCustomCheckbox title="PROMS" name="promsDate" isChecked={selected.promsDate} onChange={this.toggleCheckbox} />
                <Col xs={6} sm={4}>
                  <PTButton className="btn btn-success btn-inverse btn-bold btn-smaller" onClick={this.toggleMultipleCheckboxes(['ordersDate', 'resultsDate', 'vitalsDate', 'diagnosesDate', 'promsDate'])}>
                    <span className="btn-text">Select All</span>
                  </PTButton>
                </Col>
              </Row>
            </div>
          </div>
          <div>
            <div className="heading">COUNT</div>
            <div className="form-group">
              <Row>
                <PTCustomCheckbox title="Orders" name="ordersCount" isChecked={selected.ordersCount} onChange={this.toggleCheckbox} />
                <PTCustomCheckbox title="Results" name="resultsCount" isChecked={selected.resultsCount} onChange={this.toggleCheckbox} />
                <PTCustomCheckbox title="Vitals" name="vitalsCount" isChecked={selected.vitalsCount} onChange={this.toggleCheckbox} />
                <PTCustomCheckbox title="Diagnosis" name="diagnosesCount" isChecked={selected.diagnosesCount} onChange={this.toggleCheckbox} />
                <PTCustomCheckbox title="PROMS" name="promsCount" isChecked={selected.promsCount} onChange={this.toggleCheckbox} />
                <Col xs={6} sm={4}>
                  <PTButton className="btn btn-success btn-inverse btn-bold btn-smaller" onClick={this.toggleMultipleCheckboxes(['ordersCount', 'resultsCount', 'vitalsCount', 'diagnosesCount', 'promsCount'])}>
                    <span className="btn-text">Select All</span>
                  </PTButton>
                </Col>
              </Row>
            </div>
          </div>
        </div>
      </div>
    )
  }
}
