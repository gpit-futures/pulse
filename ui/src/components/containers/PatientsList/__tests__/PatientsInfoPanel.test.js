import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';

import PatientsInfoPanel from '../header/PatientsInfoPanel';

Enzyme.configure({ adapter: new Adapter() });
const selectedColumns = {
  name: true,
  gender: true,
  dateOfBirth: true,
  address: true,
  id: true,
  ordersDate: true,
  resultsDate: true,
  vitalsDate: false,
  diagnosesDate: false,
  ordersCount: true,
  resultsCount: true,
  vitalsCount: false,
  diagnosesCount: false,
  viewPatientNavigation: true,
};
const onColumnsSelected = () => console.log('onColumnsSelected worked');

describe('Component <PatientsInfoPanel />', () => {
  it('should renders correctly shallow', () => {
    const patientsInfoPanel = shallow(
      <PatientsInfoPanel
        selectedColumns={selectedColumns}
        onColumnsSelected={onColumnsSelected}
      />
    ).dive();
    expect(patientsInfoPanel.state().isFilterInputVisible).toEqual(false);
    expect(patientsInfoPanel.state().isPatientInfoPanelVisible).toEqual(false);
    patientsInfoPanel.find('[name="address"]').simulate('click');
    patientsInfoPanel.find('.btn-success').at(1).simulate('click');
    expect(patientsInfoPanel).toMatchSnapshot();
  });
});

