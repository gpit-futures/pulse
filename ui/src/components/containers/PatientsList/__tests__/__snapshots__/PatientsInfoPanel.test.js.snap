// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <PatientsInfoPanel /> should renders correctly shallow 1`] = `
<div
  className="dropdown-menu dropdown-menu-panel dropdown-menu-right dropdown-menu-summary dropdown-menu-patients"
>
  <div
    className="form-group-wrapper"
  >
    <div>
      <div
        className="heading"
      >
        PATIENT INFO
      </div>
      <div
        className="form-group"
      >
        <Row
          bsClass="row"
          componentClass="div"
        >
          <PTCustomCheckbox
            disabled={true}
            isChecked={true}
            name="name"
            title="Name"
          />
          <PTCustomCheckbox
            isChecked={true}
            name="address"
            onChange={[Function]}
            title="Address"
          />
          <PTCustomCheckbox
            disabled={true}
            isChecked={true}
            name="dateOfBirth"
            title="Born"
          />
          <PTCustomCheckbox
            disabled={true}
            isChecked={true}
            name="gender"
            title="Gender"
          />
          <PTCustomCheckbox
            isChecked={true}
            name="id"
            onChange={[Function]}
            title="NHS No."
          />
          <Col
            bsClass="col"
            componentClass="div"
            sm={4}
            xs={6}
          >
            <PTButton
              className="btn btn-success btn-inverse btn-bold btn-smaller"
              onClick={[Function]}
            >
              <span
                className="btn-text"
              >
                Select All
              </span>
            </PTButton>
          </Col>
        </Row>
      </div>
    </div>
    <div>
      <div
        className="heading"
      >
        DATE / TIME
      </div>
      <div
        className="form-group"
      >
        <Row
          bsClass="row"
          componentClass="div"
        >
          <PTCustomCheckbox
            isChecked={true}
            name="ordersDate"
            onChange={[Function]}
            title="Orders"
          />
          <PTCustomCheckbox
            isChecked={true}
            name="resultsDate"
            onChange={[Function]}
            title="Results"
          />
          <PTCustomCheckbox
            isChecked={true}
            name="vitalsDate"
            onChange={[Function]}
            title="Vitals"
          />
          <PTCustomCheckbox
            isChecked={true}
            name="diagnosesDate"
            onChange={[Function]}
            title="Diagnosis"
          />
          <PTCustomCheckbox
            isChecked={true}
            name="promsDate"
            onChange={[Function]}
            title="PROMS"
          />
          <Col
            bsClass="col"
            componentClass="div"
            sm={4}
            xs={6}
          >
            <PTButton
              className="btn btn-success btn-inverse btn-bold btn-smaller"
              onClick={[Function]}
            >
              <span
                className="btn-text"
              >
                Select All
              </span>
            </PTButton>
          </Col>
        </Row>
      </div>
    </div>
    <div>
      <div
        className="heading"
      >
        COUNT
      </div>
      <div
        className="form-group"
      >
        <Row
          bsClass="row"
          componentClass="div"
        >
          <PTCustomCheckbox
            isChecked={true}
            name="ordersCount"
            onChange={[Function]}
            title="Orders"
          />
          <PTCustomCheckbox
            isChecked={true}
            name="resultsCount"
            onChange={[Function]}
            title="Results"
          />
          <PTCustomCheckbox
            isChecked={false}
            name="vitalsCount"
            onChange={[Function]}
            title="Vitals"
          />
          <PTCustomCheckbox
            isChecked={false}
            name="diagnosesCount"
            onChange={[Function]}
            title="Diagnosis"
          />
          <PTCustomCheckbox
            name="promsCount"
            onChange={[Function]}
            title="PROMS"
          />
          <Col
            bsClass="col"
            componentClass="div"
            sm={4}
            xs={6}
          >
            <PTButton
              className="btn btn-success btn-inverse btn-bold btn-smaller"
              onClick={[Function]}
            >
              <span
                className="btn-text"
              >
                Select All
              </span>
            </PTButton>
          </Col>
        </Row>
      </div>
    </div>
  </div>
</div>
`;
